internal/server/storage/drivers/driver_linstor.go

```diff
@@ -0,0 +1,232 @@
+// we can't request the version from the controller unless we know its endpoint,
+// which is not the case when we first startup the server and load all available
+// drivers.
+var linstorVersion string
+var linstorLoaded bool
+
+// linstor represents the Linstor storage driver.
+type linstor struct {
+	common
+}
+
+func (d *linstor) load() error {
+	// Register the patches.
+	d.patches = map[string]func() error{
+		"storage_lvm_skipactivation":                         nil,
+		"storage_missing_snapshot_records":                   nil,
+		"storage_delete_old_snapshot_records":                nil,
+		"storage_zfs_drop_block_volume_filesystem_extension": nil,
+		"storage_prefix_bucket_names_with_project":           nil,
+	}
+
+	// Done if previously loaded.
+	if linstorLoaded {
+		return nil
+	}
+
+	// TODO: validate that the linstor-satellite service is running
+
+	// Validate the DRBD minimum version. The module should be already loaded by the
+	// Linstor satellite service.
+	ver, err := d.drbdVersion()
+	if err != nil {
+		return fmt.Errorf("Could not load Linstor driver: %w", err)
+	}
+
+	if ver.Major < 9 {
+		return fmt.Errorf("Could not load Linstor driver: Linstor requires DRBD version 9.0 to be loaded, got: %s", ver)
+	}
+
+	linstorLoaded = true
+
+	return nil
+}
+
+// isRemote returns true indicating this driver uses remote storage.
+func (d *linstor) isRemote() bool {
+	return true
+}
+
+// Validate checks that all provide keys are supported and that no conflicting or missing configuration is present.
+func (d *linstor) Validate(config map[string]string) error {
+	rules := map[string]func(value string) error{
+		LinstorResourceGroupNameConfigKey:        validate.IsAny,
+		LinstorResourceGroupPlaceCountConfigKey:  validate.Optional(validate.IsUint32),
+		LinstorResourceGroupStoragePoolConfigKey: validate.IsAny,
+		"volatile.pool.pristine":                 validate.IsAny,
+	}
+
+	return d.validatePool(config, rules, d.commonVolumeRules())
+}
+
+// FillConfig populates the storage pool's configuration file with the default values.
+func (d *linstor) FillConfig() error {
+	if d.config[LinstorResourceGroupNameConfigKey] == "" {
+		d.config[LinstorResourceGroupNameConfigKey] = LinstorDefaultResourceGroupName
+	}
+
+	if d.config[LinstorResourceGroupPlaceCountConfigKey] == "" {
+		d.config[LinstorResourceGroupPlaceCountConfigKey] = LinstorDefaultResourceGroupPlaceCount
+	}
+
+	return nil
+}
+
+// Create is called during storage pool creation.
+func (d *linstor) Create() error {
+	d.logger.Debug("Creating Linstor storage pool")
+	rev := revert.New()
+	defer rev.Fail()
+
+	// Track the initial source
+	d.config["volatile.initial_source"] = d.config["source"]
+
+	// Fill default config values
+	err := d.FillConfig()
+	if err != nil {
+		return fmt.Errorf("Could not create Linstor storage pool: %w", err)
+	}
+
+	// If a source is provided, override the resource group name
+	if d.config["source"] != "" {
```

Comment:
So this part of the code leads to a very tricky bug in the tests: *every* other storage driver sets the `source` key, and it’s tested in the clustering tests. I suggest doing the following to both make the tests work and be consistent with the other drivers:

```go
if d.config["source"] == "" {
    d.config["source"] = d.config[LinstorResourceGroupNameConfigKey]
} else {
    d.config[LinstorResourceGroupNameConfigKey] = d.config["source"]
}
```

To be fair, I’m not convinced we need LinstorResourceGroupNameConfigKey when we already have `source`…

Reply:
This is a blocking point for my patch of the test suite, so I’m waiting for this comment to be handled

Reply:
I've added an extra check for the consistency between the two config options. Also, if both `LinstorResourceGroupNameConfigKey` and `source` are not set, we fallback to the storage pool name and also set the `source` key to that value. This should match the ceph driver behavior 1:1